torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -0,0 +1,118 @@
+    : ownerId_(ownerId), rrefId_(rrefId), forkId_(forkId) {}
+
+at::IValue RRefForkData::toIValue() const {
+  std::vector<at::IValue> ivalues = {
+      (int64_t) ownerId_,
+      rrefId_.toIValue(),
+      forkId_.toIValue()
+  };
+
+  return c10::ivalue::Tuple::create(std::move(ivalues));
+}
+
+RRefForkData RRefForkData::fromIValue(at::IValue&& ivalue) {
+  auto ivalues = ivalue.toTuple()->elements();
+
+  TORCH_CHECK(ivalues.size() == 3, "Constructing RRefForkData from ivalue "
+      "expects a GenericList of 3 elements, but got ", ivalues.size());
+
+  int64_t ownerId = ivalues[0].toInt();
+  TORCH_CHECK(ownerId < std::numeric_limits<worker_id_t>::max(),
+      "RRefId createdOn out of range, got ", ownerId);
+
+  RRefId rrefId = RRefId::fromIValue(std::move(ivalues[1]));
+  ForkId forkId = ForkId::fromIValue(std::move(ivalues[2]));
+
+  return RRefForkData(ownerId, rrefId, forkId);
+}
+
+//////////////////////////////  RRef  /////////////////////////////////////
+
+RRef::RRef(worker_id_t ownerId, const RRefId& rrefId)
+    : ownerId_(ownerId), rrefId_(rrefId) {}
+
+worker_id_t RRef::owner() const {
+  return ownerId_;
+}
+
+const RRefId& RRef::id() const {
+  return rrefId_;
+}
+
+at::IValue RRef::fork() const {
+  return RRefForkData(
+      ownerId_, rrefId_, RRefContext::getInstance()->genRRefId()
+  ).toIValue();
+  // NB: does not support sharing RRefs between users
+  // TODO: notify the owner
+}
+
+//////////////////////////  UserRRef  /////////////////////////////////////
+
+
+UserRRef::UserRRef(
+    worker_id_t ownerId, const RRefId& rrefId, const ForkId& forkId)
+    : RRef(ownerId, rrefId), forkId_(forkId) {
+  AT_ASSERT(!(forkId_ == rrefId_),
+      "User RRef's fork ID should not be the same as its rref Id");
+  if (RRefContext::getInstance()->getWorkerId() == rrefId_.createdOn_) {
+    // creator user, notify owner.
+    auto& agent = RRefContext::getInstance()->agent();
+    agent->send(
+        agent->getWorkerId(ownerId_),
+        ScriptRRefCreate(
+            RRefForkData(ownerId_, rrefId_, forkId_).toIValue()
+        ).toMessage());
+  } else {
+    AT_ERROR("Does not support sharing RRefs between users yet");
+  }
+}
+
+UserRRef::~UserRRef() {
+  auto& ctx = RRefContext::getInstance();
+  if (ctx->getWorkerId() != ownerId_) {
+    ctx->agent()->send(
+        ctx->agent()->getWorkerId(ownerId_),
+        ScriptRRefDelete(
+            RRefForkData(ownerId_, rrefId_, forkId_).toIValue()
+        ).toMessage());
+  }
+}
+
+const ForkId& UserRRef::forkId() const {
+  return forkId_;
+}
+
+bool UserRRef::isOwner() const {
+  return false;
+}
+
+IValue UserRRef::toHere() {
+  auto& agent = RRefContext::getInstance()->agent();
+  std::shared_ptr<FutureMessage> fm =
+      agent->send(
+          agent->getWorkerId(ownerId_),
+          ScriptRRefFetch(id().toIValue()).toMessage()
+      );
+  auto srv = ScriptRRefValue::fromMessage(fm->wait());
+  return srv.value();
```

Comment:
Is it possible to enforce move semantics here? Not sure if this will trigger a copy.

Reply:
Adding move here gives me "warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]". 

I modified the return type of `srv.value()` to `const&` to explicitly avoid additional copy.

Reply:
It is possible that returning as value implies move semantics here, given that `srv` is also local.

Reply:
A ref will be invalid after return, since srv is local.

Reply:
> A ref will be invalid after return, since srv is local.

Oh, I mean `srv.value()`'s return type is `IValue&`, but `toHere()`'s return type is still `IValue`. So that there will only be one copy.